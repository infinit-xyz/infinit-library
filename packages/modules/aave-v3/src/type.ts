import { Address } from 'viem'

export type AaveV3Registry = {
  poolAddressProviderRegistry?: Address
  poolAddressesProvider?: Address
  liquidationLogic?: Address
  supplyLogic?: Address
  eModeLogic?: Address
  flashLoanLogic?: Address
  borrowLogic?: Address
  bridgeLogic?: Address
  poolLogic?: Address
  configuratorLogic?: Address
  reservesSetupHelper?: Address
  poolProxy?: Address
  lendingPools?: Record<string, LendingPool>
  walletBalanceProvider?: Address
  uiIncentiveDataProvider?: Address
  uiPoolDataProviderV3?: Address
  aaveProtocolDataProvider?: Address
  l2PoolImpl?: Address
  poolConfiguratorImpl?: Address
  aclManager?: Address
  aaveOracle?: Address
  aaveEcosystemReserveV2Proxy?: Address
  aaveEcosystemReserveController?: Address
  aaveEcosystemReserveV2Impl?: Address
  reserveInterestRateStrategies?: Record<string, Address>
  wrappedTokenGatewayV3?: Address
  l2Encoder?: Address
  aTokenImpl?: Address
  delegationAwareATokenImpl?: Address
  stableDebtTokenImpl?: Address
  variableDebtTokenImpl?: Address
  poolConfiguratorProxy?: Address
  rewardsControllerImpl?: Address
  rewardsControllerProxy?: Address
  pullRewardsTransferStrategy?: Address
  emissionManager?: Address
  aggregatorApi3Adapters?: Record<string, Address>
  aggregatorPythAdapters?: Record<string, Address>
  aggregatorBandAdapters?: Record<string, Address>
}

export type LendingPool = {
  underlyingToken: Address
  interestRateStrategy: Address
  aToken: Address
  stableDebtToken: Address
  variableDebtToken: Address
}

export type Modify<T, R> = Omit<T, keyof R> & R
