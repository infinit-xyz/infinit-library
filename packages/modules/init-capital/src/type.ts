import { Address } from 'viem'

export type InitCapitalRegistry = {
  proxyAdmin?: Address
  accessControlManager?: Address
  api3ProxyOracleReaderProxy?: Address
  initOracleProxy?: Address
  configProxy?: Address
  liqIncentiveCalculatorProxy?: Address
  posManagerProxy?: Address
  initCoreProxy?: Address
  riskManagerProxy?: Address
  lsdApi3ProxyOracleReaderProxy?: Address
  pythOracleReaderProxy?: Address
  moneyMarketHookProxy?: Address
  initLens?: Address
  api3ProxyOracleReaderImpl?: Address
  initOracleImpl?: Address
  configImpl?: Address
  liqIncentiveCalculatorImpl?: Address
  posManagerImpl?: Address
  initCoreImpl?: Address
  riskManagerImpl?: Address
  lsdApi3ProxyOracleReaderImpl?: Address
  pythOracleReaderImpl?: Address
  moneyMarketHookImpl?: Address
  lendingPoolImpl?: Address
  irms?: Record<string, Address>
  lendingPools?: Record<string, LendingPools>
}

export type LendingPools = {
  underlyingToken: Address
  lendingPool: Address
  irm: Address
}
