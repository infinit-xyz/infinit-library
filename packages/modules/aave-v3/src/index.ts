export { AaveV3Registry as Registry } from '@/src/type'
export { actions } from '@actions/index'
