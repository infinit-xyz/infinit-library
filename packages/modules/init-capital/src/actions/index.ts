import { InfinitActionRecord } from '@infinit-xyz/core'

import { DeployApi3ProxyOracleReaderAction, DeployApi3ProxyOracleReaderActionParamsSchema } from '@actions/deployApi3ProxyOracleReader'
import { DeployDoubleSlopeIRMsAction, DeployDoubleSlopeIRMsActionParamsSchema } from '@actions/deployDoubleSlopeIRMs'
import { DeployInitCapitalAction, DeployInitCapitalActionParamsSchema } from '@actions/deployInitCapital'
import { DeployPythOracleReaderAction, DeployPythOracleReaderActionParamsSchema } from '@actions/deployPythOracleReader'
import { SetApi3ProxyOracleReaderTokensInfoAction, SetApi3ProxyOracleReaderTokensInfoActionParamsSchema } from '@actions/setApi3ProxyOracleReaderTokensInfo'
import { SetIrmAction, SetIrmActionParamsSchema } from '@actions/setIrm'
import { SetMaxPriceDeviationAction, SetMaxPriceDeviationActionParamsSchema } from '@actions/setMaxPriceDeviation'
import { SetModeStatusAction, SetModeStatusActionParamsSchema } from '@actions/setModeStatus'
import { SetOracleAction, SetOracleActionParamsSchema } from '@actions/setOracle'
import { SetPoolConfigAction, SetPoolConfigParamsSchema } from '@actions/setPoolConfig'
import {
  SetPythOracleReaderTokensInfoAction,
  SetPythOracleReaderTokensInfoActionParamsSchema,
} from '@actions/setPythOracleReaderTokensInfo'

export const actions = {
  init: {
    name: 'Deploy INIT Capital',
    actionClassName: DeployInitCapitalAction.name,
    paramsSchema: DeployInitCapitalActionParamsSchema,
    signers: ['deployer', 'accessControlManagerOwner'],
  },
  deployDoubleSlopeIRMs: {
    name: 'Deploy Double Slope IRMs',
    actionClassName: DeployDoubleSlopeIRMsAction.name,
    paramsSchema: DeployDoubleSlopeIRMsActionParamsSchema,
    signers: ['deployer'],
  },
  setIrm: {
    name: 'Set IRM',
    actionClassName: SetIrmAction.name,
    paramsSchema: SetIrmActionParamsSchema,
    signers: ['guardian'],
  },
  setModeStatus: {
    name: 'Set Mode Status',
    actionClassName: SetModeStatusAction.name,
    paramsSchema: SetModeStatusActionParamsSchema,
    signers: ['guardian'],
  },
  setPoolConfig: {
    name: 'Set Pool Config',
    actionClassName: SetPoolConfigAction.name,
    paramsSchema: SetPoolConfigParamsSchema,
    signers: ['guardian'],
  },
  setOracle: {
    name: 'Set Oracle',
    actionClassName: SetOracleAction.name,
    paramsSchema: SetOracleActionParamsSchema,
    signers: ['governor'],
  },
  setMaxPriceDeviations: {
    name: 'Set Max Price Deviation',
    actionClassName: SetMaxPriceDeviationAction.name,
    paramsSchema: SetMaxPriceDeviationActionParamsSchema,
    signers: ['governor'],
  },
  deployApi3ProxyOracleReader: {
    name: 'Deploy Api3 Proxy Oracle Reader',
    actionClassName: DeployApi3ProxyOracleReaderAction.name,
    paramsSchema: DeployApi3ProxyOracleReaderActionParamsSchema,
    signers: ['deployer'],
  },
  setApi3ProxyOracleReaderTokensInfo: {
    name: 'Set Api3 Proxy Oracle Reader Tokens Info Action',
    actionClassName: SetApi3ProxyOracleReaderTokensInfoAction.name,
    paramsSchema: SetApi3ProxyOracleReaderTokensInfoActionParamsSchema,
    signers: ['governor'],
  },
  deployPythOracleReader: {
    name: 'Deploy Pyth Oracle Reader',
    actionClassName: DeployPythOracleReaderAction.name,
    paramsSchema: DeployPythOracleReaderActionParamsSchema,
    signers: ['deployer'],
  },
  setPythOracleReaderTokensInfo: {
    name: 'Set Pyth Oracle Reader Tokens Info',
    actionClassName: SetPythOracleReaderTokensInfoAction.name,
    paramsSchema: SetPythOracleReaderTokensInfoActionParamsSchema,
    signers: ['governor'],
  },
} satisfies InfinitActionRecord

export {
  DeployDoubleSlopeIRMsAction,
  DeployInitCapitalAction,
  DeployPythOracleReaderAction,
  SetIrmAction,
  SetMaxPriceDeviationAction,
  SetModeStatusAction,
  SetOracleAction,
  SetPoolConfigAction,
  DeployApi3ProxyOracleReaderAction,
  SetApi3ProxyOracleReaderTokensInfoAction,
}
