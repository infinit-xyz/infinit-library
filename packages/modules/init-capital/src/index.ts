export { InitCapitalRegistry as Registry } from '@/src/type'
export { actions } from '@actions/index'
