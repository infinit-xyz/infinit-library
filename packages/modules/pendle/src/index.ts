export { PendleV3Registry as Registry } from '@/src/type'
export { actions } from '@actions/index'
