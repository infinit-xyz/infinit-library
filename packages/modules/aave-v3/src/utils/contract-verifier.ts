import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { AaveV3Registry, LendingPool } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class AaveV3ContractVerifier extends BaseContractVerifier<AaveV3Registry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: AaveV3Registry): Promise<ContractInfo[]> {
    const {
      lendingPools,
      reserveInterestRateStrategies,
      poolProxy,
      poolConfiguratorProxy,
      rewardsControllerProxy,
      api3Adapters,
      pythAdapters,
      bandAdapters,
      ...contractToAddressRegistry
    } = registry

    let contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (registry.poolAddressesProvider) {
      if (poolProxy) {
        contracts.push({
          address: poolProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
      if (poolConfiguratorProxy) {
        contracts.push({
          address: poolConfiguratorProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
    }

    if (lendingPools && registry.poolConfiguratorProxy) {
      const lendingPoolContractInfos = await this.getLendingPoolContractInfos(registry.poolConfiguratorProxy, lendingPools)
      contracts = contracts.concat(lendingPoolContractInfos)
    }

    if (reserveInterestRateStrategies) {
      const reserveInterestRateStrategyContractInfos = await this.getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies)
      contracts = contracts.concat(reserveInterestRateStrategyContractInfos)
    }

    if (api3Adapters) {
      const api3AdaptersContractInfos = await this.getAdapterContractInfos(api3Adapters)
      contracts = contracts.concat(api3AdaptersContractInfos)
    }

    if (pythAdapters) {
      const pythAdaptersContractInfos = await this.getAdapterContractInfos(pythAdapters)
      contracts = contracts.concat(pythAdaptersContractInfos)
    }

    if (bandAdapters) {
      const bandAdaptersContractInfos = await this.getAdapterContractInfos(bandAdapters)
      contracts = contracts.concat(bandAdaptersContractInfos)
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }

  private async getLendingPoolContractInfos(poolConfiguratorProxy: Address, lendingPools: Record<string, LendingPool>) {
    const contracts: ContractInfo[] = []
    for (const info of Object.values(lendingPools)) {
      for (const [contractName, address] of Object.entries(info)) {
        if (contractName === 'underlyingToken') continue
        contracts.push({
          address,
          constructorArgs: contractName !== 'interestRateStrategy' ? [poolConfiguratorProxy] : undefined,
        })
      }
    }
    return contracts
  }

  private async getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(reserveInterestRateStrategies)) {
      contracts.push({ address })
    }
    return contracts
  }

  private async getAdapterContractInfos(adapters: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(adapters)) {
      contracts.push({ address })
    }
    return contracts
  }
}
