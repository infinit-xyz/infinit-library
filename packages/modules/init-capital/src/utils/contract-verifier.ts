import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { InitCapitalRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<InitCapitalRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: InitCapitalRegistry): Promise<ContractInfo[]> {
    const { lendingPools, irms, ...contractToAddressRegistry } = registry
    const contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (lendingPools) {
      for (const { lendingPool, irm } of Object.values(lendingPools)) {
        contracts.push({ address: lendingPool })
        contracts.push({ address: irm })
      }
    }

    if (irms) {
      for (const address of Object.values(irms)) {
        contracts.push({ address })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
