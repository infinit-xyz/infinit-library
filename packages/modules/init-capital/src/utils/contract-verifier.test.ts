import { describe, expect, test } from 'vitest'

import { createPublicClient, http } from 'viem'
import { holesky } from 'viem/chains'

import { TokenContractVerifier } from './contract-verifier'
import { InitCapitalRegistry } from '@/src/type'

// Create a subclass to expose the protected method for testing
class TestTokenContractVerifier extends TokenContractVerifier {
  public async testGetContracts(registry: InitCapitalRegistry) {
    return this.getContracts(registry)
  }
}

describe('TokenContractVerifier.ts', () => {
  const registry: InitCapitalRegistry = {
    accessControlManager: '0x90dd5250fd06b9e6e3d048caf7f26da609cb67cc',
    proxyAdmin: '0xba3c51fc0ff32e26fe3a6bc5f001d933bdee9dd3',
    initOracleImpl: '0xcc34062115ac818ae567daec69c29e3d8e16f1ec',
    configImpl: '0x62153519c210d21f1b67de11cf60d6f467190707',
    liqIncentiveCalculatorImpl: '0x93d027ecabf0b383f61cfad54d7d8fcae7972d33',
    posManagerImpl: '0x0e03350d11f24839373262143b46d8d68e5e6649',
    initOracleProxy: '0x03701c609ea55bfe68ad06fc36760cb25317edba',
    configProxy: '0xcc9f3144f1e57d4e3be528442452912ffcaa7b3c',
    liqIncentiveCalculatorProxy: '0x2e407758a16fa3f7d464ba6146c3cc7f1854a089',
    posManagerProxy: '0x9674f70c5ceb61f990977d325abf2c0201a4c520',
    initCoreImpl: '0x4fd715125cd4e604a89130a8d573db91b086cb60',
    initCoreProxy: '0x58f5a2711c7464b950361529ca81713b35d487b1',
    feeVault: '0xdeb76598edce92ae77f2d4f88542ed2c91b8de82',
    riskManagerImpl: '0x263a5ad450bd51cea3c818ba87c80608cb283508',
    lendingPoolImpl: '0x391d0aa2efaaddc7ccb7f0dc2a1e9336719085c8',
    moneyMarketHookImpl: '0x5de9ff054084071ca111f35180cb44c2e413be92',
    riskManagerProxy: '0x5433f79e82c136c9c2dc2f48d229627f7920acaf',
    moneyMarketHookProxy: '0xe1fc830bf20308cbba2b248c543e305b979a64ec',
    initLens: '0x3cfdf9646dbc385e47dc07869626ea36be7ba3a2',
    irms: {
      StablecoinIRM: '0x9a213f53334279c128c37da962e5472ecd90554f',
      MajorcoinIRM: '0xc55a98c1b3b0942883bb37df9716bea42d7d5009',
      MemecoinIRM: '0x12604a5b388a1e1834693bfe94dddf81a60b56a2',
      LrtcoinIRM: '0xd516492bb58f07bc91c972dccb2df654653d4d33',
    },
    api3ProxyOracleReaderImpl: '0xd23bf69104cc640e68ebee83b9833d6db6f220e6',
    api3ProxyOracleReaderProxy: '0x737b13202150ad597b72f77553b0ae30bb0c69e8',
    pythOracleReaderImpl: '0xec22b69767b5edd55b7dc6a4b6fd9ba4d0b3d320',
    pythOracleReaderProxy: '0xce33621c656f68fc35a461b7c9b09d90b9a1d548',
  }

  test('getContracts: should match snapshot', async () => {
    const verifier = new TestTokenContractVerifier(
      await createPublicClient({
        chain: holesky,
        transport: http('https://1rpc.io/holesky'),
      }),
      {
        apiKey: process.env.HOLESKY_ETHERSCAN_API_KEY ?? '',
        apiUrl: 'https://api-holesky.etherscan.io/api',
        url: 'https://holesky.etherscan.io',
      },
    )
    expect(verifier.testGetContracts(registry)).resolves.toMatchSnapshot()
  })
})
