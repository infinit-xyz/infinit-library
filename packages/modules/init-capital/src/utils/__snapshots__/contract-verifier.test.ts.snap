// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TokenContractVerifier.ts > getContracts: should match snapshot 1`] = `
[
  {
    "address": "0x90dd5250fd06b9e6e3d048caf7f26da609cb67cc",
  },
  {
    "address": "0xba3c51fc0ff32e26fe3a6bc5f001d933bdee9dd3",
  },
  {
    "address": "0xcc34062115ac818ae567daec69c29e3d8e16f1ec",
  },
  {
    "address": "0x62153519c210d21f1b67de11cf60d6f467190707",
  },
  {
    "address": "0x93d027ecabf0b383f61cfad54d7d8fcae7972d33",
  },
  {
    "address": "0x0e03350d11f24839373262143b46d8d68e5e6649",
  },
  {
    "address": "0x03701c609ea55bfe68ad06fc36760cb25317edba",
  },
  {
    "address": "0xcc9f3144f1e57d4e3be528442452912ffcaa7b3c",
  },
  {
    "address": "0x2e407758a16fa3f7d464ba6146c3cc7f1854a089",
  },
  {
    "address": "0x9674f70c5ceb61f990977d325abf2c0201a4c520",
  },
  {
    "address": "0x4fd715125cd4e604a89130a8d573db91b086cb60",
  },
  {
    "address": "0x58f5a2711c7464b950361529ca81713b35d487b1",
  },
  {
    "address": "0xdeb76598edce92ae77f2d4f88542ed2c91b8de82",
  },
  {
    "address": "0x263a5ad450bd51cea3c818ba87c80608cb283508",
  },
  {
    "address": "0x391d0aa2efaaddc7ccb7f0dc2a1e9336719085c8",
  },
  {
    "address": "0x5de9ff054084071ca111f35180cb44c2e413be92",
  },
  {
    "address": "0x5433f79e82c136c9c2dc2f48d229627f7920acaf",
  },
  {
    "address": "0xe1fc830bf20308cbba2b248c543e305b979a64ec",
  },
  {
    "address": "0x3cfdf9646dbc385e47dc07869626ea36be7ba3a2",
  },
  {
    "address": "0xd23bf69104cc640e68ebee83b9833d6db6f220e6",
  },
  {
    "address": "0x737b13202150ad597b72f77553b0ae30bb0c69e8",
  },
  {
    "address": "0xec22b69767b5edd55b7dc6a4b6fd9ba4d0b3d320",
  },
  {
    "address": "0xce33621c656f68fc35a461b7c9b09d90b9a1d548",
  },
  {
    "address": "0x9a213f53334279c128c37da962e5472ecd90554f",
  },
  {
    "address": "0xc55a98c1b3b0942883bb37df9716bea42d7d5009",
  },
  {
    "address": "0x12604a5b388a1e1834693bfe94dddf81a60b56a2",
  },
  {
    "address": "0xd516492bb58f07bc91c972dccb2df654653d4d33",
  },
]
`;
